#12 Github Action CI 서버 구축

iamabear09 opened the pull request:
### CI in Github Action

1. CI 서버의 설정 값을 `application.yml`에 주입할 수 있도록 설정
   - local 환경과 CI 서버 환경 간 설정파일 분리
2. CI 서버 workflow
   1. Repository source code를 checkout 한다.
   2. Repository 안에 존재하는 docker-compose 파일을 통해 MySQL container를 up 시킨다.
   3. Java & Gradle Setup 하고 gradlew 를 사용해 build 한다. ( build task안에 test가 존재한다. )

### Docker Compose
1. MySQL의 root password를 그대로 노출하지 않고, 환경변수를 통해 주입할 수 있도록 설정한다.


### 변경 사항 및 피드백
- Local 과 CI 환경 모두 비밀 값을 사용할 필요가 없다. 즉, 그냥 하드 코딩해도 문제될게 없다.
   - CI용 workflow와 배포 용 workflow가 다르다. (생각해보면 당연하다.) 

- application.yml 즉, 기본 설정을 Local과 CI환경에서 사용할 수 있도록 했다. 
   - 만약 운영 환경이라면, prod profile 설정을 통해 운영 설정을 적용할 수 있다.

- gradlew 실행권한은 윈도우에서 push를 하기 때문에 실행 권한이 없는 상대로 push 된다.

iamabear09 committed:
#12 feature: CI server를 위한 application.yml 수정 및 추가

- loca test용 application-dev.yml 추가
- default 설정은 환경변수를 통해 DB와 연결하도록 설정

iamabear09 committed:
#12 feature: github action CI workflow

iamabear09 committed:
#12 rafactor: make gradlew executable

iamabear09 committed:
#12 refactor: workflow step 순서 변경 & EOF 추가

iamabear09 commented on monitoring/src/main/resources/application.yml:
```diff
@@ -0,0 +1,12 @@
+spring:
+  datasource:
+    url: "${DB_URL}"
+    username: "${DB_USER}"
+    password: "${DB_PW}"
```
`"${ }"`이러한 형태로 환경변수를 사용하는 것이 가장 안전한 형태이다.

- [`"${ }"` 와 ${ } 차이점](https://github.com/f-lab-edu/self-monitoring/issues/12#issuecomment-1935179201)

iamabear09 committed:
#12 refactor: github action에 event trigger PR 추가

f-lab-lyan commented on .github/workflows/continuous-integration.yml:
```diff
@@ -0,0 +1,38 @@
+  build:
+    runs-on: ubuntu-latest
+    env:
+      DB_URL: ${{ secrets.DB_URL }}
```
CI 환경은 어차피 테스트 환경이어서 이런 정보들을 노출해도 괜찮다고 생각합니다. (물론 아주 단순한 형태의 비밀번호가 더 좋겠죠? :-))

f-lab-lyan commented on .github/workflows/continuous-integration.yml:
```diff
@@ -0,0 +1,38 @@
+      - name: Setup Gradle
+        uses: gradle/actions/setup-gradle@v3
+
+      - name: Making gradlew script executable
```
https://may9noy.tistory.com/160

f-lab-lyan commented on .github/workflows/continuous-integration.yml:
```diff
@@ -0,0 +1,38 @@
+on: [pull_request, push, workflow_dispatch]
+
+jobs:
+  build:
```
For PR, `test`
For Push, `test & build`

f-lab-lyan commented on .github/workflows/continuous-integration.yml:
```diff
@@ -0,0 +1,38 @@
+
+
+      - name: Test and Build with Gradle Wrapper
+        run: ./gradlew build
```
```suggestion
        run: ./gradlew test
```

f-lab-lyan commented on .github/workflows/continuous-integration.yml:
```diff
@@ -0,0 +1,38 @@
+      - name: Making gradlew script executable
+        run: chmod +x ./gradlew
+
+
```
```suggestion
```

iamabear09 committed:
#12 refactor: CI 환경과 Local 환경 통일

- yml 기본 설정을 Local & CI 환경으로  사용
- production 환경용 yml 파일 생성

iamabear09 committed:
#12 refactor: github action docker-compose version 문제로 인해 action 사용

- github action에서 docker-compose up 시 --wait flag 사용 불가능
- 따라서 action을 사용해서 --wait flag 사용

iamabear09 commented on gradlew:
github `./gradlew` 실행 권한 변경 
1. `git update-index --chmod=+x gradlew` 
2. 퍼미션 확인 `git ls-tree head 755` 로 나오면 정상

iamabear09 commented on monitoring/docker-compose.yml:
```diff
@@ -0,0 +1,19 @@
+      MYSQL_ROOT_PASSWORD: 1234
+    ports:
+      - "3306:3306"
+    healthcheck:
+      test: ["CMD", "mysqladmin", "ping", "-h", "localhost", "-u", "root", "-p1234"]
+      interval: 10s
+      timeout: 10s
+      retries: 5
```
`docker-compose up --wait` 사용 시 `healthcheck`가 이용된다.

f-lab-lyan reviewed (approved):
고생하셨습니다.
